src/obj.rs

```diff
@@ -316,56 +283,70 @@ impl<'a> Parser<'a> {
   ///
   /// Be careful while using this function, especially in recursive parsing as it might end up with
   /// non-linear parsing.
-  fn try<P, T>(&mut self, parse: P) -> Option<T> where P: FnOnce(&mut Self) -> Result<T, ParseError> {
+  fn try<P, T>(&mut self, parse: P) -> Option<T>
+  where
+    P: FnOnce(&mut Self) -> Result<T, ParseError>,
+  {
     let mut tried = self.clone();
 
     match parse(&mut tried) {
       Ok(r) => {
         *self = tried;
         Some(r)
-      },
-      Err(_) => None
+      }
+      Err(_) => None,
     }
   }
 
   /// Possibly skips over some newlines.
   fn zero_or_more_newlines(&mut self) {
     loop {
-      match sliced(&self.peek()) {
-        None       => break,
-        Some("\n") => {},
-        Some(_)    => break,
+      match self.peek() {
+        None => break,
+        Some("\n") => {}
+        Some(_) => break,
       }
       self.advance();
     }
   }
 
   /// Parse just a constant string.
   fn parse_tag(&mut self, tag: &str) -> Result<(), ParseError> {
-    match sliced(&self.next()) {
-      None =>
-        return self.error(format!("Expected `{}` but got end of input.", tag)),
+    match self.next() {
+      None => return self.error(format!("Expected `{}` but got end of input.", tag)),
       Some(s) => {
         if s != tag {
           return self.error(format!("Expected `{}` but got {}.", tag, s));
         }
       }
     }
 
-    return Ok(())
+    return Ok(());
+  }
+
+  fn parse_tag_or_eof(&mut self, tag: &str) -> Result<(), ParseError> {
+    match self.next() {
```

Comment:
You can go with:

```rust
self.next()
  .flat_map(|s| {
    if is != tag {
      Some(self.error(format!("Expected `{}` or EOF but got {}.", tag, s)))
    } else {
      None
    }
  }).ok_or(()).transpose()
```

Reply:
Did something slightly different, but i think it's easier to read